### PR TITLE
Support changing git config through `app.ini`, use `diff.algorithm=histogram` by default

### LINKED ISSUE #13454: Support signed pushes
Git supports signing pushes since 2.2.0, we should enable it server side if git is at least that version as it's a backwards-compatible feature. Essentially we need to configure each repo or git globally with:

```ini
[receive]
    advertisePushOptions = true
    certNonceSeed = "<uniquerandomstring>"
```

Maybe the UI can also indicate push signatures, but I guess that can come later.

`certNonceSeed` could be set to a hash derived from `security.SECRET_KEY`.

https://people.kernel.org/monsieuricon/signed-git-pushes
https://github.com/git/git/blob/7f7ebe054af6d831b999d6c2241b9227c4e4e08d/Documentation/RelNotes/2.2.0.txt#L81-L87

### LINKED ISSUE #14697: app.ini option to allow clone filters
## Description

Currently to allow clone filters (upload pack filters), one have to manually set the option via `git config`, as described on [the docs](https://docs.gitea.io/en-us/clone-filters/).

To be better, there should be option on `app.ini` named `ALLOW_CLONE_FILTERS` (or `ALLOW_UPLOAD_PACK_FILTERS`) which toggle git config `allowpack.togglefilter` globally, defaulted to false (which disallow filters). To allow filters, set the option to true. 

This option can be overridden on per-repo basis.

Note that clone filters requires Git version 2.22 or later. The option above must be hardcoded to false for instances running Git versions lower than the required.

### LINKED ISSUE #23255: Default to `histogram` diff algorithm
### Feature Description

Compelling research exists to suggest that the `histogram` diff algorithm is much superior to `myers`. This makes diffs smaller and easier to read in various edge cases. If `histogram` is not available, `patience` is close but less performant.

https://link.springer.com/article/10.1007/s10664-019-09772-z

If this request is denied, could you please advise how to alter a server setup to use this? i.e., which `.gitconfig` file should I change?

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by wxiaoguang**
Close #13454 , Close #23255, Close #14697 (and maybe more related issues)

Many users have the requirement to customize the git config. This PR introduces an easy way: put the options in Gitea's app.ini `[git.config]`, then the config options will be applied to git config.

And it can support more flexible default config values, eg: now `diff.algorithm=histogram` by default. According to: https://stackoverflow.com/a/32367597/4754037 , `histogram diff` is efficient and doesn't like to cause server-side problems.

### CODE REVIEW

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -682,6 +682,28 @@ LEVEL = Info
...
+;MIRROR = 300
+;CLONE = 300
+;PULL = 300
+;GC = 60
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Git Reflog timeout in days
+;[git.reflog]
+;ENABLED = true
+;EXPIRATION = 90
```
Comment (delvh) [inline custom/conf/app.example.ini:699]: I think we can merge this section into `[git.config]`.
If I remember that PR correctly, it did nothing else than forward these settings to the git config.

Comment (delvh) [reply custom/conf/app.example.ini:699]: The only real change in that PR was that it enables the reflog by default instead of disabling it.

Comment (delvh) [reply custom/conf/app.example.ini:699]: I mean… it makes the PR breaking, but I like that a lot more than carrying on the technical debt by explicitly handling it separately.

Comment (wxiaoguang) [reply custom/conf/app.example.ini:699]: Yup, it's done by #22451 , but it had been backported .... so we can't just drop the `[git.reflog]`, it's another breaking ....

Comment (delvh) [reply custom/conf/app.example.ini:699]: But that's exactly what I mean.
Otherwise this PR isn't breaking.

Comment (wxiaoguang) [reply custom/conf/app.example.ini:699]: So I think we can keep this PR as non-breaking, and "break" the `[git.reflog]` in a separate PR

Comment (delvh) [reply custom/conf/app.example.ini:699]: Pre-1.20.0-rc0?

**Review by delvh — APPROVED** [review]

File: custom/conf/app.example.ini (hunk 1 above)
Comment (wxiaoguang) [reply custom/conf/app.example.ini:699]: Do you mean to "break" `[git.reflog]` in 1.20 release? If you agree, I can do it in the following PR.

Comment (delvh) [reply custom/conf/app.example.ini:699]: Yes, that's what I meant.
I want to get rid of that technical debt as fast as possible.

**Comment by delvh** [comment]
I've relabeled the PR now as I think this change
a) has a huge enough impact to be classified as a feature
b) Is future-proof enough that I doubt we will break it at any point
c) is technically new functionality, and not the improvement of existing functionality

File: custom/conf/app.example.ini [hunk 2]
```diff
@@ -682,6 +682,29 @@ LEVEL = Info
...
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Git config options
+;; This section only does "set" config, a removed config key from this section won't be removed from git config automatically
+;[git.config]
+;diff.algorithm = histogram
+;other.config-key = ...
```
Comment (silverwind) [inline custom/conf/app.example.ini]: This example is the not default, would remove. You can outline the example in the comment above instead.

Comment (wxiaoguang) [reply custom/conf/app.example.ini]: Can you provide a suggested change?

**Comment by silverwind** [comment]
To really solve https://github.com/go-gitea/gitea/issues/13454, it would be nice to add an example somewhere in the docs how to enable signed pushes. Not necessarily in the cheat sheet. Nonce is just a random string, like

```
    certNonceSeed = "<uniquerandomstring>"
```

**Comment by wxiaoguang** [comment]
> To really solve #13454, it would be nice to add an example somewhere in the docs how to enable signed pushes. Not necessarily in the cheat sheet. Nonce is just a random string, like
> 
> ```
>     certNonceSeed = "<uniquerandomstring>"
> ```

Feel free to edit on this PR directly ~~~ (I am not sure I can 100% write a matched example config 😁  )

File: custom/conf/app.example.ini (hunk 2 above)
Comment (silverwind) [inline custom/conf/app.example.ini]:
```suggestion
;; This section only does "set" config, a removed config key from this section won't be removed from git config automatically. The format is `some.configKey = value`.
;[git.config]
;diff.algorithm = histogram
```

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 3]
```diff
@@ -1068,6 +1063,19 @@ Default templates for project boards:
...
+- `ENABLED`: **true** Set to true to enable Git to write changes to reflogs in each repo.
+- `EXPIRATION`: **90** Reflog entry lifetime, in days. Entries are removed opportunistically by Git.
+
+### Git - Config options (`git.config`)
+
+The key/value pairs in this section will be used as git config.
+This section only does "set" config, a removed config key from this section won't be removed from git config automatically.
+
+- `diff.algorithm`: **histogram**
+- `other.config-key`: ...
```
Comment (silverwind) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]:
```suggestion
This section only does "set" config, a removed config key from this section won't be removed from git config automatically. The format is `some.configKey = value`.

- `diff.algorithm`: **histogram**
```

File: custom/conf/app.example.ini (hunk 2 above)
Comment (silverwind) [reply custom/conf/app.example.ini]: Done below.

Comment (delvh) [reply custom/conf/app.example.ini]:
```suggestion
;; This section only adds to the git config, you cannot remove a key by not specifying it. The format is `some.configKey = value`.
;[git.config]
;diff.algorithm = histogram
```

**Review by lunny — APPROVED** [review]

**Review by silverwind — DISMISSED** [review]
I will add some docs regarding signed pushes, blocking until then.

**Comment by silverwind** [comment]
Question: When adding further directives, does the user need to specify the defaults like `diff.algorithm` again? E.g. would adding signed pushes have to be

````ini
[git.config]
diff.algorithm = histogram
receive.advertisePushOptions = true
certNonceSeed = randomstring
````

or

````ini
[git.config]
receive.advertisePushOptions = true
certNonceSeed = randomstring
````

?

**Comment by silverwind** [comment]
Signed push docs added in https://github.com/go-gitea/gitea/pull/24860/commits/938d64595fd8bf783f35f172c527ed25d992e721.

File: docs/content/doc/administration/customizing-gitea.en-us.md [hunk 4]
```diff
@@ -282,6 +282,22 @@ Place custom files in corresponding sub-folder under `custom/options`.
...
+
+### Enabling signed git pushes
+
+To enable signed git pushes, set these two options:
+
+````
+[git.config]
+receive.advertisePushOptions = true
+receive.certNonceSeed = <randomstring>
+````
```
Comment (KN4CK3R) [inline docs/content/doc/administration/customizing-gitea.en-us.md]:
````suggestion
```ini
[git.config]
receive.advertisePushOptions = true
receive.certNonceSeed = <randomstring>
```
````

Comment (KN4CK3R) [reply docs/content/doc/administration/customizing-gitea.en-us.md]: hm, not how it should look like... 😄 

Change the 4 ` to 3.

Comment (delvh) [reply docs/content/doc/administration/customizing-gitea.en-us.md]: I edited it for you

File: docs/content/doc/administration/customizing-gitea.en-us.md [hunk 5]
```diff
@@ -282,6 +282,22 @@ Place custom files in corresponding sub-folder under `custom/options`.
...
+
+To enable signed git pushes, set these two options:
+
+```ini
+[git.config]
+receive.advertisePushOptions = true
+receive.certNonceSeed = <randomstring>
+```
+
+`certNonceSeed` should be set to a sufficiently random string and be kept secret.
```
Comment (silverwind) [inline docs/content/doc/administration/customizing-gitea.en-us.md]:
```suggestion
`certNonceSeed` should be set to a random string and be kept secret.
```

**Review by KN4CK3R — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> Question: When adding further directives, does the user need to specify the defaults like `diff.algorithm` again?

No. It's covered by the test:

image

**Pull request merged**